FR #1082: Clarify requirements to cover all invoiced charges in FOCUS datasets (#1321)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/columns/invoiceid.md

# Invoice ID

> An Invoice ID is a provider-assigned identifier for an invoice encapsulating all [*charges*](#glossary:charge) in the corresponding [*billing period*](#glossary:billing-period) for a given [*billing account*](#glossary:billing-account). Invoices are commonly used for scenarios like tracking billing transactions, facilitating payment processes and for performing invoice reconciliation between *charges* and billing periods.

Reviewer: ```suggestion
An Invoice ID is a provider-assigned identifier for an invoice encapsulating all [*charges*](#glossary:charge) in the corresponding [*billing period*](#glossary:billing-period) for a given [*billing account*](#glossary:billing-account). Invoices are commonly used for scenarios like tracking billing transactions, facilitating payment processes and for performing invoice reconciliation between *charges* and billing periods.
```

Putting 'all' here feels like the wrong messaging @thecloudman @rileyjenk @shawnalpay. Providers can provide multiple invoices for the same billing period and billing account for various reasons. In some cases, AWS Support, for example, comes in a separate invoice, even though you have an original invoice for the billing account and period. There are possibly other cases like purchases or corrections, where you have multiple invoices/invoice ids.

A possible improvement would be to say the sum of all charges provided by invoice issuers that provide invoice ids for a given BillingAccount and BillingPeriod should match the sum of all the invoices associated with the same scope.

Author: agreed, reverting the language to give space for multiple invoices within a single billing period.

The InvoiceId column adheres to the following requirements:

* InvoiceId is RECOMMENDED to be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* InvoiceId MUST be of type String.
* InvoiceId MUST conform to [StringHandling](#stringhandling) requirements.

> * The sum of BilledCost values for all rows sharing the same InvoiceId MUST match the total payable amount on the corresponding invoice, unless exceptions defined in the InvoiceHandling attribute apply.

Reviewer 2: ```suggestion
* The sum of BilledCost for a given [InvoiceId](#invoiceid) MUST match the sum of the payable amount provided in the corresponding invoice with the same id generated by the [InvoiceIssuer](#InvoiceIssuer).
```

Personally, I prefer the original version as specified in section 1.2 under the BilledCost column for several reasons:

- While unlikely, it's still possible for two different InvoiceIssuers to generate invoices with the same Invoice ID. Therefore, I think it's valuable to keep the phrasing: "with the same id generated by the InvoiceIssuer."

- Phrasing differences: In other normative requirements, we haven't explicitly mentioned "values" (e.g., BilledCost values), as it's generally implied that the ColumnId refers to the value itself.

- Question: Do the exceptions listed in the attribute apply to the entire invoice, even the one related to the withheld charges? If that's the case, than there is no exception in this requirement.

Positioning of the requirement: This requirement should be placed last — anyway not before the nullability-related requirements.

## Description

> The provider-assigned identifier for an invoice encapsulating all *charges* in the corresponding billing period for a given billing account.

Reviewer 3: I don't think we can make this change. Billing periods may be covered by multiple invoices. Or at least, we'd need to say something like, "...that were billed on the invoice identified by the invoice ID." (But in a much more graceful way 🙃)

Author: I reverted back to some or all, but gave the additional clarification that the full set of charges (can span across multiple invoices) need to be represented within the FOCUS dataset. Let me know what you think

Reviewer 4: I struggle to reconcile the "some or all" part here with line 21 that states:

> All costs that appear on an invoice MUST be present in the *FOCUS dataset*.

In any case, the Description section should strive to be as short as possible.  Any nuance you add like this should be captured elsewhere in the markdown.  Thus, here, I would suggest watering it down further to:

```suggestion
The provider-assigned identifier for an invoice encapsulating *charges* in the corresponding billing period for a given billing account.
```

...And then adding that nuance either up at the top of the document, or in the normative requirements.

Reviewer 3: I like this. Simple and safe.

Author: I've applied these recommended changes in the latest commit, though Github wouldn't let me commit the suggestion directly. as it's related to an outdated comment.
